### PR TITLE
Namespace the option parser

### PR DESCRIPTION
**Pull request by wthrowe**
Everything except the first two commits will be squashed together, but it may be easier to review split up.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Options/Options.hpp [hunk 1]
```diff
@@ -26,6 +26,8 @@ class Node;
 }  // namespace YAML
 /// \endcond
 
+/// Utilities for parsing input files.
+namespace Options {
```
Comment (nilsvu) [inline src/Options/Options.hpp:30]: How about making this namespace lower-case (`options`)? Not sure how annoying the regex-replace would be to make this change though.. what do you think?

File: src/Evolution/DiscontinuousGalerkin/Limiters/Tags.hpp [hunk 2]
```diff
@@ -25,7 +25,9 @@ struct LimiterGroup {
  */
 template <typename LimiterType>
 struct Limiter {
-  static std::string name() noexcept { return option_name<LimiterType>(); }
+  static std::string name() noexcept {
+    return Options::option_name<LimiterType>();
```
Comment (nilsvu) [inline src/Evolution/DiscontinuousGalerkin/Limiters/Tags.hpp]: Perhaps just `options::name()`?

File: src/Options/OptionsDetails.hpp [hunk 3]
```diff
@@ -184,14 +185,15 @@ struct print_impl<Tag, OptionList,
...
+           << (MakeString{} << std::boolalpha << Tag::default_value());
       }
     }
     if constexpr (has_lower_bound<Tag>::value) {
-      ss << "\n    min=" << Tag::lower_bound();
+      ss << "\n    min=" << (MakeString{} << Tag::lower_bound());
     }
     if constexpr (has_upper_bound<Tag>::value) {
-      ss << "\n    max=" << Tag::upper_bound();
+      ss << "\n    max=" << (MakeString{} << Tag::upper_bound());
```
Comment (nilsvu) [inline src/Options/OptionsDetails.hpp:196]: For my understanding, why is this change needed?

Comment (nilsdeppe) [reply src/Options/OptionsDetails.hpp:196]: I would guess it tricks ADL into looking into the global namespace. I'm unsure if a `using ::operator<<;` would work here because that might depend on include order -_-

**Review by nilsdeppe — DISMISSED** [review]
This looks great! Thanks for doing this @wthrowe!

Some Thoughts on Nils's comments:
- I have no preference between `Options` and `options` in general. I would vote to "leave as-is" since that's less work
- If the `Options::name()` change is easy to do, I agree that it would be nice and also consistent with naming changes in other commits in this PR. However, definitely non-binding :)

I vote we ignore clang-tidy on this PR because it inevitably had to touch a lot of files and it's not fair to @wthrowe to have to fix all the clang-tidy things, especially in a PR that does so much.

I'm happy for this to be squashed :)

**Comment by nilsvu** [comment]
I agree with @nilsdeppe's comments, and please squash directly if you make any changes.

File: src/Options/OptionsDetails.hpp (hunk 3 above)
Comment (wthrowe) [reply src/Options/OptionsDetails.hpp:196]: > For my understanding, why is this change needed?

I don't know.  I don't understand how overload resolution interacts with namespaces (except for the basic rule that the namespaces containing the argument types are searched).  The original code stopped working, and this worked.

`using ::operator<<;` fails to compile on some files because it finds nothing to import.

**Review by wthrowe — COMMENTED** [review]
> I have no preference between `Options` and `options` in general. I would vote to "leave as-is" since that's less work

I like `options` better, but it conflicts with all the `using options = tmpl::list<...>` everywhere.

> If the `Options::name()` change is easy to do, I agree that it would be nice and also consistent with naming changes in other commits in this PR. However, definitely non-binding :)

Done.

> I vote we ignore clang-tidy on this PR because it inevitably had to touch a lot of files and it's not fair to @wthrowe to have to fix all the clang-tidy things, especially in a PR that does so much.

I fixed the one in `Test_TestCreation.cpp` since that's actual option parsing code.  I left the others.

File: src/Options/OptionsDetails.hpp (hunk 3 above)
Comment (nilsdeppe) [reply src/Options/OptionsDetails.hpp:196]: I think the rule is that since `MakeString` is in the global namespace and `std::BLAH` is in the `std` namespace `operator<<` is searched for in both the global and `std` namespace. Since we can't add `operator<<` for `std` things into the `std`namespace, using `MakeString` gets the compiler to search in the global namespace, which is where we have `operator<<` defined for the STL containers. I think the reason `using operator<<` fails is because it's eager while the `MakeString` trick causes the compiler to search for the stream operator when the function is used, not when it is defined :roll_eyes:

**Review by nilsdeppe — APPROVED** [review]

**Review by nilsvu — APPROVED** [review]
> > I have no preference between `Options` and `options` in general. I would vote to "leave as-is" since that's less work
> 
> I like `options` better, but it conflicts with all the `using options = tmpl::list<...>` everywhere.

Oh right, that's annoying..
